### ISSUE TITLE
Lint/AmbiguousRange and Style/RedundantParentheses cause infinite loop

### ISSUE THREAD

**Issue by ccutrer** [rubocop/rubocop#12436]
## Expected behavior

When parentheses are required to remove ambiguity from a range, Style/RedundantParentheses should not remove them.

## Steps to reproduce the problem

```ruby
# frozen_string_literal: true

a = {}
a[:description] = a[:description][0...a[:description].index("<!--BB ASSIGNMENT INTERNALS: SKIP REST-->") || -1] if a[:description]
```

```
 % rubocop --only Lint/AmbiguousRange,Style/RedundantParentheses -A test.rb --debug
For /Users/cody/src: Default configuration from /Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/cody/src/test.rb
W

Offenses:

test.rb:4:39: W: [Corrected] Lint/AmbiguousRange: Wrap complex range boundaries with parentheses to avoid ambiguity.
a[:description] = a[:description][0...a[:description].index("<!--BB ASSIGNMENT INTERNALS: SKIP REST-->") || -1] if a[:description]
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test.rb:4:39: C: [Corrected] Style/RedundantParentheses: Don't use parentheses around a logical expression.
a[:description] = a[:description][0...(a[:description].index("<!--BB ASSIGNMENT INTERNALS: SKIP REST-->") || -1)] if a[:description]
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0 files inspected, 2 offenses detected, 2 offenses corrected
Infinite loop detected in /Users/cody/src/test.rb and caused by Lint/AmbiguousRange -> Style/RedundantParentheses
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:332:in `check_for_infinite_loop'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:315:in `block in iterate_until_no_changes'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:314:in `loop'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:314:in `iterate_until_no_changes'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:283:in `do_inspection_loop'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:164:in `block in file_offenses'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:189:in `file_offense_cache'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:163:in `file_offenses'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:154:in `process_file'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:135:in `block in each_inspected_file'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `each'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `reduce'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `each_inspected_file'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:120:in `inspect_files'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/runner.rb:73:in `run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/command.rb:11:in `run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli/environment.rb:18:in `run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli.rb:118:in `run_command'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli.rb:125:in `execute_runners'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli.rb:51:in `block in run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli.rb:77:in `profile_if_needed'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/lib/rubocop/cli.rb:43:in `run'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/exe/rubocop:19:in `block in <top (required)>'
/Users/cody/.rubies/ruby-3.1.3/lib/ruby/3.1.0/benchmark.rb:311:in `realtime'
/Users/cody/.gem/ruby/3.1.3/gems/rubocop-1.58.0/exe/rubocop:19:in `<top (required)>'
/Users/cody/.gem/ruby/3.1.3/bin/rubocop:25:in `load'
/Users/cody/.gem/ruby/3.1.3/bin/rubocop:25:in `<main>'
Finished in 0.08233999999902153 seconds
```

## RuboCop version

```
% rubocop -V
1.58.0 (using Parser 3.2.2.4, rubocop-ast 1.30.0, running on ruby 3.1.3) [arm64-darwin21]
```
